motoko/encrypted-notes-dapp/src/encrypted_notes_rust/src/lib.rs

```diff
@@ -0,0 +1,591 @@
+///
+/// See https://smartcontracts.org/docs/developers-guide/concepts/canisters-code.html#query-update
+
+/// Reflects the [caller]'s identity by returning (a future of) its principal. 
+/// Useful for debugging.
+#[update(name = "whoami")]
+fn whoami() -> String {
+    caller_api().to_string()
+}
+
+/// The following invariant is preserved by [register_device].
+///
+/// All the functions of this canister's public API are available only to 
+/// registered users, with the exception of [register_device] and [whoami].
+///
+/// See also: [is_user_registered]
+fn users_invariant() -> bool {
+    let num_users_with_notes = NOTES_BY_USER.with(|notes_ref| notes_ref.borrow().keys().len() as u128);
+    let num_users_with_store = USER_KEYS.with(|user_keys_ref| user_keys_ref.borrow().keys().len() as u128);
+    num_users_with_notes == num_users_with_store
+}
+
+/// Returns the current number of users.
+/// Panics if [users_invariant] is violated.
+fn user_count() -> usize {
+    assert!(users_invariant());
+    NOTES_BY_USER.with(|notes_ref| notes_ref.borrow().keys().len())
+}
+
+/// Check if this user has been registered
+/// Note: [register_device] must be each user's very first update call.
+/// See also: [users_invariant]
+fn is_user_registered(principal: Principal) -> bool {
+    USER_KEYS.with(|user_keys_ref| user_keys_ref.borrow().contains_key(&principal))
+}
+
+/// Check that a note identifier is sane. This is needed since we use finite-
+/// precision integers (`u128`).
+fn is_id_sane(id: u128) -> bool {
+    MAX_NOTES_PER_USER.with(|max_notes_per_user| id < (*max_notes_per_user as u128) * (user_count() as u128))
+}
+
+/// Returns (a future of) this [caller]'s notes.
+/// Panics: 
+///     [caller] is the anonymous identity
+///     [caller] is not a registered user
+#[update(name = "get_notes")]
+fn get_notes() -> Vec<EncryptedNote> {
+    let user = caller();
+    assert!(is_user_registered(user));
+    let user_str = user.to_string();
+    NOTES_BY_USER.with(|notes_ref| {
+        notes_ref
+            .borrow()
+            .get(&user_str)
+            .cloned()
+            .unwrap_or_default()
+    })
+}
+
+/// Delete this [caller]'s note with given id. If none of the 
+/// existing notes have this id, do nothing. 
+/// [id]: the id of the note to be deleted
+///
+/// Returns: 
+///      Future of unit
+/// Panics: 
+///      [caller] is the anonymous identity
+///      [caller] is not a registered user
+///      [id] is unreasonable; see [is_id_sane]
+#[update(name = "delete_note")]
+fn delete_note(note_id: u128) {
+    let user = caller();
+    assert!(is_user_registered(user));
+    assert!(is_id_sane(note_id));
+
+    let user_str = user.to_string();
+    // shared ownership borrowing
+    NOTES_BY_USER.with(|notes_ref| {
+        let mut writer = notes_ref.borrow_mut();
+        if let Some(v) = writer.get_mut(&user_str) {
+            v.retain(|item| item.id != note_id);
+        }
```

Comment:
don't we want to reduce the counter in this case?

Reply:
Not in the current design, because the counter is used as a source of unique note ids.